libflux/src/semantic/types.rs

```diff
@@ -116,9 +217,136 @@ impl fmt::Display for MonoType {
     }
 }
 
+impl Substitutable for MonoType {
+    fn apply(self, sub: &Subst) -> Self {
+        match self {
+            MonoType::Bool => MonoType::Bool,
+            MonoType::Int => MonoType::Int,
+            MonoType::Uint => MonoType::Uint,
+            MonoType::Float => MonoType::Float,
+            MonoType::String => MonoType::String,
+            MonoType::Duration => MonoType::Duration,
+            MonoType::Time => MonoType::Time,
+            MonoType::Regexp => MonoType::Regexp,
+            MonoType::Var(tvr) => tvr.apply(sub),
+            MonoType::Arr(arr) => MonoType::Arr(Box::new(arr.apply(sub))),
+            MonoType::Row(obj) => MonoType::Row(Box::new(obj.apply(sub))),
+            MonoType::Fun(fun) => MonoType::Fun(Box::new(fun.apply(sub))),
+        }
+    }
+}
+
+impl MonoType {
+    fn unify(self, with: Self, cons: Const) -> Result {
+        match (self, with) {
+            (MonoType::Bool, MonoType::Bool)
+            | (MonoType::Int, MonoType::Int)
+            | (MonoType::Uint, MonoType::Uint)
+            | (MonoType::Float, MonoType::Float)
+            | (MonoType::String, MonoType::String)
+            | (MonoType::Duration, MonoType::Duration)
+            | (MonoType::Time, MonoType::Time)
+            | (MonoType::Regexp, MonoType::Regexp) => Ok((Subst::empty(), cons)),
+            (MonoType::Var(tv), t) => tv.unify(t, cons),
+            (t, MonoType::Var(tv)) => tv.unify(t, cons),
+            (MonoType::Arr(t), MonoType::Arr(s)) => t.unify(*s, cons),
+            (MonoType::Row(t), MonoType::Row(s)) => t.unify(*s, cons),
+            (MonoType::Fun(t), MonoType::Fun(s)) => t.unify(*s, cons),
+            (t, with) => Err(Error::cannot_unify(t, with)),
+        }
+    }
+
+    fn constrain(self, with: HashSet<Kind>, cons: Const) -> Result {
+        match self {
+            MonoType::Bool => self
+                .satisfies_at_most(maplit::hashset![Kind::Equatable, Kind::Nullable,])(
+                with, cons
+            ),
+            MonoType::Int => self.satisfies_at_most(maplit::hashset![
+                Kind::Addable,
+                Kind::Subtractable,
+                Kind::Divisible,
+                Kind::Comparable,
+                Kind::Equatable,
+                Kind::Nullable
+            ])(with, cons),
```

Comment:
It seems like every time you constrain something a new hashset gets created. Could we store those constant hashsets in something generated once for all?

Reply:
This could be handled more efficiently using nested match expressions. This ways is nice because all of the constraint and error logic is in a single closure. Let's optimize this later.